scipy/spatial/src/distance_metrics.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <cmath>
+#include "views.h"
+
+struct MinkowskiDistance {
+    double p_;
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
+        T p = p_;
+        for (int64_t i = 0; i < x.shape[0]; ++i) {
```

Comment:
My last remaining concern is the discussion on int64_t @sturlamolden brought up on int64_t:

@peterbell10: While I see using an int could cause overflow, could you explain how using intptr_t could introduce an overflow?

Usually npy_intp / intptr_t is supposed to be sufficient for indexing any numpy array, and used throughout the entire numpy codebase.

If the concern is intptr_t is not C++-like, then I looked the counter part in pybind11 (unchecked) uses ssize_t, which is somewhat of an stl favorite and similar to intptr_t:

https://github.com/pybind/pybind11/blob/e08a58111dbea38d667b209f7543864d51a3b185/include/pybind11/numpy.h#L364

Reply:
The mathematically ”correct” solution is to use ```ssize_t```, however it is a POSIX datatype and not a standard C++ datatype. 

The argument in favor of ```ssize_t``` is that the size of the address space and the maximum segment size might be different, though modern 64 processors tend to use a flat address space. This is the reason CPython ```Py_ssize_t``` instead of ```Py_intptr_t``` for indexing. 

The argument in favor of ```intptr_t``` is that it is a standard C++ datatype, whilst  ```ssizet_t``` is only required to exist on a POSIX compliant system.

The choice between  ```ssize_t``` and ```intptr_t``` might seem a bit pedantic though. NumPy and SciPy certainly uses ```intptr_t``` which is the more portable of the two.

Reply:
It is certainly the case that ```ssize_t``` and ```intptr_t``` can overflow. However, fixing offsets to 64 bits does not solve this problem universally. It still exists on 64 bit systems and on a 128 bit system it is only made worse. This is more a design problem in NumPy that we have everywhere.

Reply:
My problem is the 32-bit index math bit, not the spelling of the type. A classic example is finding the midpoint: `(low + high) / 2` can overflow even though `low` and `high` always fit in an int.

I don't think any of the calculations here are at risk of overflow, so I'll concede.

Reply:
```low + (high - low)/2```

Reply:
It's not impossible. However, the obvious way is not correct and the correct way is non-obvious. A bug like this existed in SciPy and it caused segfaults for large input arrays. Yet it had worked without issue for 20 years.

>using 64 bit ints does not prevent an expression like (low+high)/2 from producing an overflow. 

True, but only you need to have 4 million terrabytes of memory before that becomes possible.

Reply:
The use of low + (high - low) / 2 is (nowadays) relatively well documented. See, e.g. https://en.wikipedia.org/wiki/Binary_search_algorithm#Implementation_issues

Reply:
> True, but you need to have 4 million terrabytes of memory before that becomes possible

That’s what I am planning for my next laptop 😁